// common/views/login/index.jsx
import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';

import FieldInput from 'common/forms/field-input';
import FieldCheckbox from 'common/forms/field-checkbox';
import Form from 'common/forms/form';
import Button from 'common/components/button';
import { makeUrl } from 'common/utils/url-helper';

class Login extends React.Component {
  static propTypes = {
    location: React.PropTypes.object,
    res: React.PropTypes.object,
    isAuthenticated: React.PropTypes.bool,
  }

  componentWillMount() {
    const { isAuthenticated, res } = this.props;

    if (isAuthenticated) {
      if (res) {
        res.redirect('/dashboard');
      }
    }
  }

  getErrors() {
    const { location } = this.props;

    if (location && location.query && location.query.errors) {
      try {
        return JSON.parse(location.query.errors);
      } catch (err) {
        return null;
      }
    }

    return null;
  }

  getValues() {
    const { location } = this.props;

    if (location && location.query && location.query.values) {
      try {
        return JSON.parse(location.query.values);
      } catch (err) {
        return null;
      }
    }

    return null;
  }

  render() {
    const errors = this.getErrors();
    const values = this.getValues();

    return (
      <div className="login">
        <Form action={makeUrl('/login')}>
          <div className="login__item">
            <FieldInput
              label="Username"
              id="username"
              name="username"
              error={
                _.get(errors, 'username', null)
              }
              defaultValue={
                _.get(values, 'username', null)
              }
            />
          </div>
          <div className="login__item">
            <FieldInput
              label="Password"
              id="password"
              name="password"
              type="password"
              error={
                _.get(errors, 'password', null)
              }
            />
          </div>
          <div className="login__item">
            <FieldCheckbox
              name="remember"
              id="remember"
              label="Remember me"
              value="1"
            />
          </div>
          <div className="login__item">
            <Button type="submit">Submit</Button>
          </div>
        </Form>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
});

export default connect(mapStateToProps)(Login);
