// common/forms/field-input/index.jsx
import React from 'react';

class FieldInput extends React.Component {
  static propTypes = {
    label: React.PropTypes.string,
    validation: React.PropTypes.string,
    id: React.PropTypes.string.isRequired,
    name: React.PropTypes.string.isRequired,
    type: React.PropTypes.string,
    value: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.number,
    ]),
    error: React.PropTypes.string,
  };

  static defaultProps = {
    type: 'text',
    value: null,
  };

  render() {
    const { id, label, type, error, ...rest } = this.props;

    return (
      <div className="field-input">
        {!!label ?
          <label className="field-input__label" htmlFor={id}>{label}</label>
        : null}

        {!!error ?
          <span className="field-input__error">{error}</span>
        : null}

        <input
          type={type}
          id={id}
          className="field-input__field"
          {...rest}
        />
      </div>
    );
  }
}

export default FieldInput;
