// common/forms/field-input-select/index.jsx
import React from 'react';

import Button from 'common/components/button';

class FieldInputSelect extends React.Component {
  static propTypes = {
    label: React.PropTypes.string,
    value: React.PropTypes.string,
    display: React.PropTypes.string,
    handleClick: React.PropTypes.func,
    error: React.PropTypes.string,
    id: React.PropTypes.string,
    onChange: React.PropTypes.func,
  };

  componentWillUpdate(nextProps) {
    const { value, onChange } = this.props;

    if (value !== nextProps.value && onChange) {
      onChange(nextProps.value);
    }
  }

  render() {
    const { label, id, error, handleClick, display, value, ...rest } = this.props;

    return (
      <div className="field-input-select">
      {!!label ?
          <label className="field-input-select__label" htmlFor={id}>{label}</label>
        : null}

        {!!error ?
          <span className="field-input-select__error">{error}</span>
        : null}

        <Button
          size="large"
          className="field-input-select__select"
          onClick={handleClick}
        >
          Select
        </Button>
        <span className="field-input-select__search-wrapper">
          <input type="text" className="field-input-select__search" value={display} readOnly />
          <input type="hidden" id={id} value={value} {...rest} />
        </span>
      </div>
    );
  }
}

export default FieldInputSelect;
