// common/reducers/deviceTypes.jsx
function deviceTypes(state = {
  index: {
    filter: null,
    status: null,
    meta: null,
    deviceTypes: null,
  },
  edit: {
    status: null,
    meta: null,
    deviceType: null,
  },
}, action) {
  switch (action.type) {
    case 'REQUEST_DEVICE_TYPES':
      return state;
    case 'RECEIVE_DEVICE_TYPES':
      return Object.assign({}, state, {
        index: {
          filter: state.index.filter,
          status: action.response.status,
          meta: action.response.meta,
          deviceTypes: action.response.data,
        },
      });
    case 'REQUEST_DEVICE_TYPE':
      return state;
    case 'RECEIVE_DEVICE_TYPE':
      return Object.assign({}, state, {
        edit: {
          status: action.response.status,
          meta: action.response.meta,
          deviceType: action.response.data,
        },
      });
    case 'CLEAR_DEVICE_TYPES_INDEX':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: null,
          meta: null,
          deviceTypes: null,
        },
      });
    case 'CLEAR_EDIT_DEVICE_TYPE':
      return Object.assign({}, state, {
        edit: {
          status: null,
          meta: null,
          deviceTypes: null,
        },
      });
    case 'FILTER_DEVICE_TYPES':
      return Object.assign({}, state, {
        index: {
          filter: action.filter,
          status: null,
          meta: null,
          users: null,
        },
      });
    case 'CLEAR_DEVICE_TYPES_FILTER':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: state.index.status,
          meta: state.index.meta,
          deviceTypes: state.index.deviceTypes,
        },
      });
    default:
      return state;
  }
}

export default deviceTypes;
