// common/reducers/devices.jsx
export function devices(state = {
  index: {
    filter: null,
    status: null,
    meta: null,
    devices: null,
  },
  edit: {
    status: null,
    meta: null,
    device: null,
  },
  deviceTypeModal: {
    visible: false,
    label: null,
    value: null,
  },
}, action) {
  switch (action.type) {
    case 'REQUEST_DEVICES':
      return state;
    case 'RECEIVE_DEVICES':
      return Object.assign({}, state, {
        index: {
          filter: state.index.filter,
          status: action.response.status,
          meta: action.response.meta,
          devices: action.response.data,
        },
      });
    case 'SHOW_DEVICE_TYPE_MODAL':
      return Object.assign({}, state, {
        deviceTypeModal: {
          visible: true,
          label: state.deviceTypeModal.label,
          value: state.deviceTypeModal.value,
        },
      });
    case 'HIDE_DEVICE_TYPE_MODAL':
      return Object.assign({}, state, {
        deviceTypeModal: {
          visible: false,
          label: state.deviceTypeModal.label,
          value: state.deviceTypeModal.value,
        },
      });
    case 'STORE_DEVICE_TYPE_VALUE':
      return Object.assign({}, state, {
        deviceTypeModal: {
          visible: false,
          label: action.label,
          value: action.value,
        },
      });
    case 'CLEAR_DEVICE_TYPE_VALUE':
      return Object.assign({}, state, {
        deviceTypeModal: {
          visible: state.deviceTypeModal.visible,
          label: null,
          value: null,
        },
      });
    case 'CLEAR_DEVICES_INDEX':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: null,
          meta: null,
          devices: null,
        },
      });
    case 'FILTER_DEVICES':
      return Object.assign({}, state, {
        index: {
          filter: action.filter,
          status: null,
          meta: null,
          devices: null,
        },
      });
    case 'CLEAR_DEVICES_FILTER':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: state.index.status,
          meta: state.index.meta,
          devices: state.index.devices,
        },
      });
    case 'REQUEST_DEVICE':
      return state;
    case 'RECEIVE_DEVICE':
      return Object.assign({}, state, {
        edit: {
          status: action.response.status,
          meta: action.response.meta,
          device: action.response.data,
        },
      });
    case 'CLEAR_EDIT_DEVICE':
      return Object.assign({}, state, {
        edit: {
          status: null,
          meta: null,
          device: null,
        },
      });
    default:
      return state;
  }
}
