// common/reducers/rules.jsx
function rules(state = {
  index: {
    filter: null,
    status: null,
    meta: null,
    rules: null,
  },
  edit: {
    status: null,
    meta: null,
    rule: null,
  },
  inputDeviceModal: {
    visible: false,
    label: null,
    value: null,
  },
}, action) {
  switch (action.type) {
    case 'REQUEST_RULES':
      return state;
    case 'RECEIVE_RULES':
      return Object.assign({}, state, {
        index: {
          filter: state.index.filter,
          status: action.response.status,
          meta: action.response.meta,
          rules: action.response.data,
        },
      });
    case 'REQUEST_RULE':
      return state;
    case 'RECEIVE_RULE':
      return Object.assign({}, state, {
        edit: {
          status: action.response.status,
          meta: action.response.meta,
          rule: action.response.data,
        },
      });
    case 'SHOW_INPUT_DEVICE_MODAL':
      return Object.assign({}, state, {
        inputDeviceModal: {
          visible: true,
          label: state.inputDeviceModal.label,
          value: state.inputDeviceModal.value,
        },
      });
    case 'HIDE_INPUT_DEVICE_MODAL':
      return Object.assign({}, state, {
        inputDeviceModal: {
          visible: false,
          label: state.inputDeviceModal.label,
          value: state.inputDeviceModal.value,
        },
      });
    case 'STORE_INPUT_DEVICE_VALUE':
      return Object.assign({}, state, {
        inputDeviceModal: {
          visible: false,
          label: action.label,
          value: action.value,
        },
      });
    case 'CLEAR_INPUT_DEVICE_VALUE':
      return Object.assign({}, state, {
        inputDeviceModal: {
          visible: false,
          label: null,
          value: null,
        },
      });
    case 'CLEAR_RULES_INDEX':
      return Object.assign({}, state, {
        index: {
          filter: state.index.filter,
          status: null,
          meta: null,
          rules: null,
        },
      });
    case 'CLEAR_RULES_FILTER':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: state.index.status,
          meta: state.index.meta,
          rules: state.index.rules,
        },
      });
    case 'CLEAR_EDIT_RULE':
      return Object.assign({}, state, {
        edit: {
          status: null,
          meta: null,
          rule: null,
        },
      });
    case 'FILTER_RULES':
      return Object.assign({}, state, {
        index: {
          filter: action.filter,
          status: null,
          meta: null,
          rules: null,
        },
      });
    default:
      return state;
  }
}

export default rules;
