// common/reducers/users.jsx
function users(state = {
  index: {
    filter: null,
    status: null,
    meta: null,
    users: null,
  },
  edit: {
    status: null,
    meta: null,
    users: null,
  },
}, action) {
  switch (action.type) {
    case 'REQUEST_USERS':
      return state;
    case 'RECEIVE_USERS':
      return Object.assign({}, state, {
        index: {
          filter: state.index.filter,
          status: action.response.status,
          meta: action.response.meta,
          users: action.response.data,
        },
      });
    case 'REQUEST_USER':
      return state;
    case 'RECEIVE_USER':
      return Object.assign({}, state, {
        edit: {
          status: action.response.status,
          meta: action.response.meta,
          user: action.response.data,
        },
      });
    case 'CLEAR_USERS_INDEX':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: null,
          meta: null,
          users: null,
        },
      });
    case 'CLEAR_EDIT_USER':
      return Object.assign({}, state, {
        edit: {
          status: null,
          meta: null,
          users: null,
        },
      });
    case 'REFRESH_USERS_INDEX':
      return Object.assign({}, state, {
        index: {
          status: null,
          meta: state.index.meta,
          users: null,
        },
      });
    case 'FILTER_USERS':
      return Object.assign({}, state, {
        index: {
          filter: action.filter,
          status: null,
          meta: null,
          users: null,
        },
      });
    case 'CLEAR_USERS_FILTER':
      return Object.assign({}, state, {
        index: {
          filter: null,
          status: state.index.status,
          meta: state.index.meta,
          users: state.index.users,
        },
      });
    default:
      return state;
  }
}

export default users;
